tokio/src/runtime/dump.rs

```diff
@@ -214,6 +216,66 @@ impl Trace {
             })
             .collect()
     }
+
+    /// Runs the function `f` in tracing mode, and returns its result along with the resulting [`Trace`].
+    ///
+    /// The passed-in function is normally a poll function. Due to the way tracing is implemented,
+    /// Tokio leaf futures might, instead of doing their actual work, do the equivalent of a
+    /// `yield_now` (returning a `Poll::Pending` and scheduling the current context for execution),
+    /// which means forward progress is only guaranteed if you eventually call your future outside of
+    /// `capture`.
+    ///
+    /// Example usage:
+    /// ```
+    /// let mut trace = None;
+    ///
+    /// // some future
+    /// let mut test_future = Box::pin(async move { tokio::task::yield_now().await; 0 });
+    ///
+    /// // trace it once
+    /// Trace::root(std::future::poll_fn(|cx| {
+    ///     if trace.is_none() {
+    ///         // make sure we only trace once
+    ///         let (res, captured) = Trace::capture(|| test_future.as_mut().poll(cx));
+    ///         trace = Some(captured);
+    ///         res
+    ///     } else {
+    ///         test_future.as_mut().poll(cx)
+    ///     }
+    /// })).await;@@
+    ///
+    /// // check that there are backtraces
+    /// assert!(!trace.unwrap().resolve_backtraces().is_empty());
```

Comment:
This example isn't so nice. I would probably do something along these lines:
```rs
// some future
let mut test_future = std::pin!(async move { tokio::task::yield_now().await; 0 });

// trace it once
let (trace, res) = Trace::root(std::future::poll_fn(|cx| {
    let (res, trace) = Trace::capture(|| test_future.as_mut().poll(cx));
    Poll::Ready((trace, res))
})).await;

let output = match res {
    Poll::Ready(output) => output,
    Poll::Pending => test_future.await,
};

println!("{trace}");
```
Perhaps we should have a function along these lines to make it easier?

Reply:
Will switch to your example. My intended use is closer to the `task_trace_self` test. I think a helper function would better fit in a separate PR.